Docs: remove time travel query page (#28646)
MaterializeInc/materialize: doc/user/content/transform-data/patterns/durable-subscriptions.md

---
title: "Durable subscription (Retain History)"

> description: "How to configure history retention to set up durable subscriptions in Materialize"

Author: Oh ... in talking with Frank, I guess there are still other losses ... not the one this feature will prevent.  So, committing your suggestion, but in the next patch, I'll just remove the "lossless" part.

> description: "How to configure history retention to set up durable subscriptions in Materialize"

Reviewer: Sounds good.

menu:
  main:
    parent: 'sql-patterns'
aliases:
  - /transform-data/patterns/time-travel-queries/
  - /transform-data/patterns/time-travel-queries/#history-retention-period
---

You can configure an object's [history retention
period](#history-retention-period) for durable subscriptions.

{{< private-preview />}}

## Durable subscriptions

Because [`SUBSCRIBE`](/sql/subscribe/) requests happen over the network, these
connections might get disrupted for both expected and unexpected reasons. To
avoid the need for re-snapshotting the data, you can:

> - Adjust the [history retention period](#history-retention-period) for the
>   objects that a subscription depends on, and then

Reviewer: ```suggestion
- Adjust the [history retention period](#history-retention-period) for the
  objects that a subscription depends on, and
```

> - Use [`AS OF`](/sql/subscribe#as-of) to pick up where you left off on
>   connection drops.

Reviewer: ```suggestion
- [Access past versions of this data](#enabling-durable-subscriptions-in-your-application)
  at specific points in time to pick up data processing where you left off.
```

Reviewer: Might be more user-friendly to structure this as a guide with two steps and/or explicitly point to each section instead of using implicit links.

## History retention period

{{< private-preview />}}

> The history retention period of an object ([sources](/sql/create-source/),
> [tables](/sql/create-table/), [materialized
> views](/sql/create-materialized-view/), and [indexes](/sql/create-index/))
> determines the duration for which historical versions of the object's underlying
> data are available for querying. By default, history retention period is unset
> for all user-defined sources, tables, materialized views, and indexes; that is,
> these objects keep track only of the most recent version of their underlying
> data. You can modify this retention period using the `RETAIN HISTORY` option.

Reviewer: ```suggestion
By default, all user-defined sources, tables, materialized views, and indexes
keep track of the most recent version of their underlying data. To gracefully
recover from connection disruptions and enable lossless, _durable
subscriptions_, you can configure the objects the subscription depends on
to **retain history**.

To configure the history retention period for an object, use the `RETAIN
HISTORY` option in its `CREATE` statement. This value can also be adjusted at
any time using the object-specific `ALTER` statement.
```

Reviewer: Editing in GH, so unpolished, but the line of thought should be more focused on the functionality (durable subscriptions) rather than the implementation detail (history retention).

> ### Behavior

Reviewer: ```suggestion
### Semantics
```

Reviewer: All of this should live somewhere else since it's reference documentation. Surfacing a comment I left in the original PR (#26623):

> Eventually, I think we'll want to extricate the SQL reference bits as a shortcode that can be applied to the low-level reference pages, and focus this page on just guidance on how to implement the time traveling patterns in practice.

Author: 100% agreed.  Yeah ... I noticed that this page seemed more SQL reference than how to implement the durable subscription using a feature of our product.

Reviewer: I meant this as a comment for a future pass, if you don't have the bandwidth right now. This is the current state, anyways, so it's not like it'll get worse. 🫠 Probably something we'll just forget to do if we don't do it now, is more the thing!

Author: Since this is the single use case for this option, I can implement the shortcode in the next patch.  That way, this page really can be about implementing durable subscription instead of trying to be about that + a reference section on the history retention feature.

    for that cluster to go up, which might require you to size up your cluster
    and consume additional compute credits.

### Clean-up
#### Best practices

> The history retention period represents the minimum amount of historical data

Reviewer: What's the motivation to cut the details on clean-up? AFAIR it was included because a user asked about how history cleanup worked.

Author: Not cut.  Just moved up to the Behavior/Semantics section 
 https://github.com/MaterializeInc/materialize/pull/28646/files#diff-cf72fc8c3a01d715512ded2b51842cf0cb083481d0065b3afc379f00230e9889R74

Author: Sounds good.  Will move it there in the next patch of this PR.

  index, consider creating a materialized view for your subscription query and

>   configure the history retention period on that materialized view instead.

Reviewer: ```suggestion
  configuring the history retention period on that materialized view instead.
```

> - Similarly, because of the increased storage costs and processing time for the
>   additional historical data on whichever object has an increased history
>   retention period, you should consider configuring history retention period on
>   the index or materialized view directly powering the subscription, rather than
>   all the way through the dependency chain from the source to the index or
>   materialized view.

Reviewer: ```suggestion
- Similarly, because of the increased storage costs and processing time for the
  additional historical data on whichever object has an increased history
  retention period, consider configuring history retention period on the index
  or materialized view directly powering the subscription, rather than all the
  way through the dependency chain from the source to the index or materialized
  view.
```

Reviewer: Nit, but "consider" vs. "you should consider" from your modification above.

> ## Set up durable subscriptions in your application

Reviewer: ```suggestion
## Enabling durable subscriptions in your application
```

Reviewer: The other same-level section is called `History retention period`. These should be consistent.

Author: Yup.  This was where the page got goofy trying to be both a procedural page about how to implement durable subscriptions but also be a placeholder for SQL reference material on History retention 😄

1. In Materialize, configure the history retention period for the object(s)
queried in the `SUBSCRIBE`. Choose a duration you expect will allow you to
recover in case of connection drops. One hour (`1h`) is a good place to start,

> though you should be mindful of the [resource utilization impact and other
> considerations](#considerations) of increasing an object's history retention
> period.

Reviewer: ```suggestion
though you should be mindful of the [impact](#considerations) of increasing an
object's history retention period.
```

   You can tweak the flush interval at which you durably record the latest
   progress timestamp, if every progress message is too frequent.

### Note about idempotency

> The procedure above recommends you buffer data in memory until receiving a

Reviewer: ```suggestion
The guidance above recommends you buffer data in memory until receiving a
```

progress message, and then persist the data and progress message `mz_timestamp`

> at the same time. This is to ensure data is processed exactly once.

Reviewer: ```suggestion
at the same time. This is to ensure data is processed **exactly once**.
```

In the case that your application crashes and you need to resume your subscription using the
persisted progress message `mz_timestamp`:
* If you were processing data in your application before persisting the subsequent progress message's `mz_timestamp`: you may end up processing duplicate data.
* If you were persisting the progress message's `mz_timestamp` before processing all the
buffered data from before that progress message: you may end up dropping some data.

> As a result, to guarantee that the data processing occurs only once after your
> application crashes is to write the progress message `mz_timestamp` and all
> buffered data **together in a single transaction**.

Reviewer: ```suggestion
As a result, to guarantee that the data processing occurs only once after your
application crashes, you must write the progress message `mz_timestamp` and all
buffered data **together in a single transaction**.
```
